### ISSUE TITLE
OrderlessQRNADE missing, plus a few minor issues

### ISSUE THREAD

**Issue by remidomingues** [MarcCote/NADE#1]
Hello,

We are currently writing a research paper including a benchmark of your algorithm, but we stumbled across a few issues:

- While running the DeepNADE with the 'QR' form, we found out that the file `deepnade/buml/NADE/OrderlessQRNADE.py` was missing from the git repository. Would it be possible to push it?

- While training the network (single-layer MoG or Laplace, or multi-layer MoG) on normalized datasets containing continuous features (e.g. abalone, yeast or shuttle from UCI), we found out that most trainings were stopped before completion due to NaN encountered in the computations. Is there a way to avoid such issues by fine-tuning a few specific network parameters?

- Eventually and after convergence, we observed that most log-likelihoods values returned for distinct data points were identical. Would you have any hint about what might be causing it?

Many thanks for your help.

**Comment by MarcCote** [comment]
Hi, I'll contact @uria as he coded those parts.

**Comment by remidomingues** [comment]
Hi, any update on your side? We are ready to launch our experiments as soon as NADE is on board.

**Comment by uria** [comment]
Hi,

I never published the QR form of NADE (based on quantile regression)  the
file should be missing, I never intended to release it. I'm sorry if it is
referenced from the other files.

Regarding NaNs:
  Yes, training these things is finicky. Make sure the data is not
real-valued but actually only takes a small set of values (common in some
UCI datasets), as NADE will learn to put more and more density on those
values and the training will diverge. That type of dataset shouldn't be
modeled with real-valued density estimators in my opinion. The usual tricks
for training these models apply: floor the variances, weight decay, adding
noise to data, smaller learning rates...

--Beni

**Comment by remidomingues** [comment]
Hello Uria, and many thanks for your answer.

The datasets we are using are real-valued, and we thought that your algorithm would be a good match for our benchmark as your paper stated that RNADE was compatible with such data. Can you confirm that the `OrderlessNADERunner` class using a `MoG` or `Laplace` form is suitable for such experiment?

We generated the following density estimation plot using your algorithm, which shows an excellent density estimation at convergence.

![image](https://user-images.githubusercontent.com/5238619/30021199-d2a8a346-9166-11e7-8bf7-7ed5a1f678fc.png)

However, for most of our datasets, the training either fails with NaNs, or most of the log-likelihoods values returned are identical for numerous distinct data points. We tried fine-tuning your algorithm (learning rate, weight decay, number of units and components, layers, form, etc...) but that didn't solve the issue.

Would you have any idea of what might be causing it, and how we could solve this issue?
A great many thanks for your help.

**Comment by remidomingues** [comment]
Hello,

We solved the problem on our side. Here are a few issues which caused the training to fail:
- We were normalizing our datasets, then split them into training, validation and testing datasets. As such, the training dataset mean was slightly deviating from 0, which was sufficient to cause an unstable training.
- The normalization function from utils does not currently handle zeros in standard deviation, i.e. when the dataset contains constant features. This was solved by `std[std == 0.0] = 1.0`.

Have a nice day.
